nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ModifyCompression.java

```diff
@@ -0,0 +1,453 @@
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        final long sizeBeforeCompression = flowFile.getSize();
+
+        String decompressionFormatValue = context.getProperty(DECOMPRESSION_FORMAT).getValue();
+        if (decompressionFormatValue.equals(COMPRESSION_FORMAT_ATTRIBUTE)) {
+            final String mimeType = flowFile.getAttribute(CoreAttributes.MIME_TYPE.key());
+            if (mimeType == null) {
+                logger.error("No {} attribute exists for {}; routing to failure", CoreAttributes.MIME_TYPE.key(), flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+
+            decompressionFormatValue = compressionFormatMimeTypeMap.get(mimeType);
+            if (decompressionFormatValue == null) {
+                logger.info("MIME Type of {} is '{}', which does not indicate a supported Decompression Format; routing to failure", flowFile, mimeType);
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+        }
+
+        String compressionFormatValue = context.getProperty(COMPRESSION_FORMAT).getValue();
+        if (compressionFormatValue.equals(COMPRESSION_FORMAT_ATTRIBUTE)) {
+            final String mimeType = flowFile.getAttribute(CoreAttributes.MIME_TYPE.key());
+            if (mimeType == null) {
+                logger.error("No {} attribute exists for {}; routing to failure", CoreAttributes.MIME_TYPE.key(), flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+                return;
+            }
+
+            compressionFormatValue = compressionFormatMimeTypeMap.get(mimeType);
+            if (compressionFormatValue == null) {
+                logger.info("MIME Type of {} is '{}', which does not indicate a supported Compression Format; routing to success without decompressing", flowFile, mimeType);
+                session.transfer(flowFile, REL_SUCCESS);
+                return;
+            }
+        }
+
+        final String decompressionFormat = decompressionFormatValue;
+        final String compressionFormat = compressionFormatValue;
+        final AtomicReference<String> mimeTypeRef = new AtomicReference<>(null);
+        final StopWatch stopWatch = new StopWatch(true);
+
+        final String inputFileExtension = getFileExtension(decompressionFormat);
+        final String outputFileExtension = getFileExtension(compressionFormat);
+
+        try {
+            flowFile = session.write(flowFile, (rawIn, rawOut) -> {
```

Comment:
With the callback method being rather large, recommend refactoring it to a separate inner class.